File: src/pixiis/services/oauth.py
"""Browser-based OAuth with local redirect capture.

Starts a temporary HTTP server on localhost, opens the user's browser to the
authorization URL, waits for the redirect callback, and returns the
token/code parameters.  Works with both query-based redirects (Google auth
code) and fragment-based redirects (Twitch implicit grant).
"""

from __future__ import annotations

import threading
import webbrowser
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import parse_qs, urlparse


# ── Fragment bridge page ────────────────────────────────────────────────────
# Twitch implicit grant puts the token in the URL *fragment* (#access_token=…)
# which is never sent to the server.  This page extracts the fragment with JS
# and forwards it as a regular query-string GET to /token on the same server.

_FRAGMENT_HTML = """\
<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Connecting...</title>
<style>
  body { font-family: system-ui, sans-serif; background: #1a1a2e; color: #e0e0e8;
         display: flex; align-items: center; justify-content: center; height: 100vh; margin: 0; }
  .card { text-align: center; padding: 2rem; }
  .ok   { color: #4ade80; font-size: 1.5rem; }
  .fail { color: #f87171; font-size: 1.5rem; }
</style></head><body>
<div class="card" id="msg">Connecting&hellip;</div>
<script>
(function() {
  var hash = window.location.hash.substring(1);
  if (!hash) {
    document.getElementById("msg").innerHTML = '<span class="fail">No token received.</span>';
    return;
  }
  var xhr = new XMLHttpRequest();
  xhr.open("GET", "/token?" + hash, true);
  xhr.onload = function() {
    document.getElementById("msg").innerHTML = '<span class="ok">Connected! You can close this tab.</span>';
  };
  xhr.onerror = function() {
    document.getElementById("msg").innerHTML = '<span class="fail">Failed to send token to app.</span>';
  };
  xhr.send();
})();
</script></body></html>
"""

_SUCCESS_HTML = """\
<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Done</title>
<style>
  body { font-family: system-ui, sans-serif; background: #1a1a2e; color: #4ade80;
         display: flex; align-items: center; justify-content: center; height: 100vh; margin: 0; }
</style></head><body>
<h2>Connected! You can close this tab.</h2>
</body></html>
"""


class OAuthFlow:
    """Browser-based OAuth with local redirect capture.

    Usage::

        flow = OAuthFlow()
        redirect_uri = f"http://localhost:{flow.port}/callback"
        auth_url = build_auth_url(redirect_uri=redirect_uri)
        flow.start(auth_url)
        result = flow.get_result(timeout=120)
        # result is a dict like {"access_token": "...", "token_type": "bearer"}
    """

    def __init__(self) -> None:
        self._result: dict[str, str] | None = None
        self._event = threading.Event()

        # Bind to a random available port on localhost
        self._server = HTTPServer(("127.0.0.1", 0), self._make_handler())
        self._thread: threading.Thread | None = None

    @property
    def port(self) -> int:
        """The port the local server is listening on."""
        return self._server.server_address[1]

    # ── public API ──────────────────────────────────────────────────────

    def start(self, auth_url: str) -> None:
        """Open the browser and start listening.  Non-blocking."""
        self._thread = threading.Thread(target=self._serve, daemon=True)
        self._thread.start()
        webbrowser.open(auth_url)

    def get_result(self, timeout: float = 120) -> dict[str, str] | None:
        """Block until the redirect is received or *timeout* seconds elapse.

        Pass ``timeout=0`` for a non-blocking check (returns immediately).
        Returns ``None`` if no result yet or on timeout.
        """
        if timeout <= 0:
            return self._result
        self._event.wait(timeout=timeout)
        return self._result

    def cancel(self) -> None:
        """Shut down the local server early."""
        self._shutdown()

    # ── internals ───────────────────────────────────────────────────────

    def _serve(self) -> None:
        """Run the HTTP server until shut down."""
        self._server.serve_forever()

    def _finish(self, params: dict[str, str]) -> None:
        """Store the result and shut down."""
        self._result = params
        self._event.set()
        # Shut down in a separate thread so the handler can return its response
        threading.Thread(target=self._shutdown, daemon=True).start()

    def _shutdown(self) -> None:
        self._server.shutdown()

    def _make_handler(self) -> type[BaseHTTPRequestHandler]:
        """Build a request handler class with a reference back to this flow."""
        flow = self

        class _Handler(BaseHTTPRequestHandler):
            def do_GET(self) -> None:  # noqa: N802
                parsed = urlparse(self.path)

                if parsed.path == "/callback":
                    qs = parse_qs(parsed.query)
                    flat = {k: v[0] for k, v in qs.items()}
                    if flat:
                        # Query-based redirect (e.g. auth-code flow)
                        self._respond(200, _SUCCESS_HTML)
                        flow._finish(flat)
                    else:
                        # No query params — probably a fragment-based redirect.
                        # Serve the JS bridge page so it can read the fragment.
                        self._respond(200, _FRAGMENT_HTML)

                elif parsed.path == "/token":
                    # Second request from the JS bridge with fragment params
                    qs = parse_qs(parsed.query)
                    flat = {k: v[0] for k, v in qs.items()}
                    self._respond(200, "ok")
                    if flat:
                        flow._finish(flat)

                else:
                    self._respond(404, "Not found")

            def _respond(self, code: int, body: str) -> None:
                self.send_response(code)
                content_type = (
                    "text/html; charset=utf-8"
                    if body.startswith("<!") or body.startswith("<h")
                    else "text/plain"
                )
                self.send_header("Content-Type", content_type)
                self.send_header("Access-Control-Allow-Origin", "*")
                encoded = body.encode("utf-8")
                self.send_header("Content-Length", str(len(encoded)))
                self.end_headers()
                self.wfile.write(encoded)

            def log_message(self, format: str, *args: object) -> None:  # noqa: A002
                # Suppress console output from the HTTP server
                pass

        return _Handler

File: src-tauri/src/services/oauth.rs
//! Local OAuth callback server.
//!
//! Mirrors `src/pixiis/services/oauth.py`:
//!
//! * Binds an axum server to `127.0.0.1:0` (random port).
//! * `GET /callback?…` with query string → finishes (auth-code grant).
//! * `GET /callback` with no query → serves the JS bridge page that reads
//!   `window.location.hash` and re-issues `GET /token?…` (Twitch implicit).
//! * `GET /token?…` → finishes.
//!
//! Once finished, the server shuts down on its own. `start()` opens the
//! browser at the supplied auth URL.

use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::{Arc, Mutex};
use std::time::Duration;

use axum::extract::{Query, State};
use axum::http::header;
use axum::response::{IntoResponse, Response};
use axum::routing::get;
use axum::Router;
use tokio::net::TcpListener;
use tokio::sync::oneshot;
use tokio::time::timeout;

use crate::error::{AppError, AppResult};

const FRAGMENT_HTML: &str = r#"<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Connecting...</title>
<style>
  body { font-family: system-ui, sans-serif; background: #1a1a2e; color: #e0e0e8;
         display: flex; align-items: center; justify-content: center; height: 100vh; margin: 0; }
  .card { text-align: center; padding: 2rem; }
  .ok   { color: #4ade80; font-size: 1.5rem; }
  .fail { color: #f87171; font-size: 1.5rem; }
</style></head><body>
<div class="card" id="msg">Connecting&hellip;</div>
<script>
(function() {
  var hash = window.location.hash.substring(1);
  if (!hash) {
    document.getElementById("msg").innerHTML = '<span class="fail">No token received.</span>';
    return;
  }
  var xhr = new XMLHttpRequest();
  xhr.open("GET", "/token?" + hash, true);
  xhr.onload = function() {
    document.getElementById("msg").innerHTML = '<span class="ok">Connected! You can close this tab.</span>';
  };
  xhr.onerror = function() {
    document.getElementById("msg").innerHTML = '<span class="fail">Failed to send token to app.</span>';
  };
  xhr.send();
})();
</script></body></html>"#;

const SUCCESS_HTML: &str = r#"<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Done</title>
<style>
  body { font-family: system-ui, sans-serif; background: #1a1a2e; color: #4ade80;
         display: flex; align-items: center; justify-content: center; height: 100vh; margin: 0; }
</style></head><body>
<h2>Connected! You can close this tab.</h2>
</body></html>"#;

/// Owns one in-flight OAuth flow.
///
/// Lifecycle: `start()` -> caller opens browser -> `wait_for_result()` blocks
/// (with timeout) until the redirect lands. Server self-terminates on success
/// or via `cancel()` for early teardown. If neither fires, a built-in
/// watchdog (10 min by default) tears it down so dropped flows don't leak.
pub struct OAuthFlow {
    port: u16,
    result_rx: Mutex<Option<oneshot::Receiver<HashMap<String, String>>>>,
    shutdown: Arc<Mutex<Option<oneshot::Sender<()>>>>,
}

const DEFAULT_WATCHDOG: Duration = Duration::from_secs(600);

#[derive(Clone)]
struct AppState {
    finish: Arc<Mutex<Option<oneshot::Sender<HashMap<String, String>>>>>,
    shutdown: Arc<Mutex<Option<oneshot::Sender<()>>>>,
}

impl OAuthFlow {
    /// Bind to a random localhost port and start serving. Returns once the
    /// listener is bound — the server itself runs in a spawned tokio task.
    pub async fn start() -> AppResult<Self> {
        let listener = TcpListener::bind(SocketAddr::from(([127, 0, 0, 1], 0)))
            .await
            .map_err(AppError::Io)?;
        let port = listener.local_addr().map_err(AppError::Io)?.port();

        let (result_tx, result_rx) = oneshot::channel();
        let (shutdown_tx, shutdown_rx) = oneshot::channel();
        let shutdown = Arc::new(Mutex::new(Some(shutdown_tx)));

        let state = AppState {
            finish: Arc::new(Mutex::new(Some(result_tx))),
            shutdown: Arc::clone(&shutdown),
        };

        let app = Router::new()
            .route("/callback", get(handle_callback))
            .route("/token", get(handle_token))
            .with_state(state);

        tokio::spawn(async move {
            let _ = axum::serve(listener, app)
                .with_graceful_shutdown(async move {
                    let _ = shutdown_rx.await;
                })
                .await;
        });

        // Watchdog: ensures the server shuts down even if the OAuthFlow is
        // dropped without anyone calling wait_for_result/cancel (e.g. the
        // user never completes the browser flow).
        {
            let watchdog = Arc::clone(&shutdown);
            tokio::spawn(async move {
                tokio::time::sleep(DEFAULT_WATCHDOG).await;
                if let Some(tx) = watchdog.lock().unwrap().take() {
                    let _ = tx.send(());
                }
            });
        }

        Ok(Self {
            port,
            result_rx: Mutex::new(Some(result_rx)),
            shutdown,
        })
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    /// Block (async) until the redirect arrives or `dur` elapses.
    pub async fn wait_for_result(&self, dur: Duration) -> AppResult<HashMap<String, String>> {
        let rx = self
            .result_rx
            .lock()
            .unwrap()
            .take()
            .ok_or_else(|| AppError::Other("oauth result already consumed".into()))?;
        let res = match timeout(dur, rx).await {
            Ok(Ok(params)) => Ok(params),
            Ok(Err(_)) => Err(AppError::Other("oauth flow cancelled".into())),
            Err(_) => Err(AppError::Other("oauth timeout".into())),
        };
        // Whether timeout, cancel, or success — make sure the server is down.
        self.cancel();
        res
    }

    /// Tear down the local server early without waiting for a callback.
    pub fn cancel(&self) {
        if let Some(tx) = self.shutdown.lock().unwrap().take() {
            let _ = tx.send(());
        }
    }
}

async fn handle_callback(
    State(state): State<AppState>,
    Query(params): Query<HashMap<String, String>>,
) -> Response {
    if params.is_empty() {
        // Fragment-based redirect — serve the JS bridge.
        return html(FRAGMENT_HTML);
    }
    finish(&state, params);
    html(SUCCESS_HTML)
}

async fn handle_token(
    State(state): State<AppState>,
    Query(params): Query<HashMap<String, String>>,
) -> Response {
    if params.is_empty() {
        return ([(header::CONTENT_TYPE, "text/plain; charset=utf-8")], "no params").into_response();
    }
    finish(&state, params);
    ([(header::CONTENT_TYPE, "text/plain; charset=utf-8")], "ok").into_response()
}

fn finish(state: &AppState, params: HashMap<String, String>) {
    if let Some(tx) = state.finish.lock().unwrap().take() {
        let _ = tx.send(params);
    }
    // Defer shutdown briefly so the response can flush before the listener
    // stops. 100ms matches the Python implementation's threading.Thread spawn.
    let shutdown = Arc::clone(&state.shutdown);
    tokio::spawn(async move {
        tokio::time::sleep(Duration::from_millis(100)).await;
        if let Some(tx) = shutdown.lock().unwrap().take() {
            let _ = tx.send(());
        }
    });
}

fn html(body: &'static str) -> Response {
    (
        [
            (header::CONTENT_TYPE, "text/html; charset=utf-8"),
            (header::ACCESS_CONTROL_ALLOW_ORIGIN, "*"),
        ],
        body,
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn callback_with_query_completes_flow() {
        let flow = OAuthFlow::start().await.unwrap();
        let port = flow.port();

        let url = format!("http://127.0.0.1:{port}/callback?code=abc&state=xyz");
        let client = reqwest::Client::new();
        let body = client
            .get(&url)
            .send()
            .await
            .unwrap()
            .text()
            .await
            .unwrap();
        assert!(body.contains("Connected"));

        let params = flow.wait_for_result(Duration::from_secs(2)).await.unwrap();
        assert_eq!(params.get("code").map(String::as_str), Some("abc"));
        assert_eq!(params.get("state").map(String::as_str), Some("xyz"));
    }

    #[tokio::test]
    async fn fragment_bridge_completes_via_token() {
        let flow = OAuthFlow::start().await.unwrap();
        let port = flow.port();
        let client = reqwest::Client::new();

        let body = client
            .get(format!("http://127.0.0.1:{port}/callback"))
            .send()
            .await
            .unwrap()
            .text()
            .await
            .unwrap();
        assert!(body.contains("/token"));

        let body = client
            .get(format!(
                "http://127.0.0.1:{port}/token?access_token=TOKEN&token_type=bearer"
            ))
            .send()
            .await
            .unwrap()
            .text()
            .await
            .unwrap();
        assert_eq!(body, "ok");

        let params = flow.wait_for_result(Duration::from_secs(2)).await.unwrap();
        assert_eq!(params.get("access_token").map(String::as_str), Some("TOKEN"));
    }

    #[tokio::test]
    async fn timeout_returns_error() {
        let flow = OAuthFlow::start().await.unwrap();
        let err = flow.wait_for_result(Duration::from_millis(50)).await;
        assert!(err.is_err());
    }
}
